### DISCUSSION TITLE
TypeError: add_label() takes exactly 2 positional arguments (3 given)

### DISCUSSION THREAD

**Discussion by Dan-Essin** [explosion/spaCy#9393] [Help: Coding & Implementations] [answered]
Hello,
I'm new to nlp and am working through a book with examples. I have reached a section called "Training our own POS-taggers".
There is a simple example that doesn't work for me. I'm still new enough at python that I cannot understand why it is complaining about the arguments to the tagger,add_label call. I can only see 2 but for some reason it things that there are 3. It would be really helpful if someone could explain this to me. 

This is the code
===============================================
import os
import gensim.models
import spacy
import spacy.lang.en.stop_words as STOP_WORDS
from gensim import corpora, models
import plac
import random
from pathlib import Path

TAG_MAP = {
    'N': {'pos': 'NOUN'},
    'V': {'pos': 'VERB'},
    'J': {'pos': 'ADJ'}
}
TRAIN_DATA = [
    ("I like green eggs", {'tags': ['N', 'V', 'J', 'N']}),
    ("Eat blue ham", {'tags': ['V', 'J', 'N']})
]

@plac.annotations(
    lang=("ISO Code of language to use", "option", "l", str),
    output_dir=("Optional output directory", "option", "o", Path),
    n_iter=("Number of training iterations", "option", "n", int))
def main(lang='en', output_dir=None, n_iter=25):
    """Main function to create a new model, set up the pipeline and train
    the tagger. In order to train the tagger with a custom tag map,
    we're creating a new Language instance with a custom vocab.
    """
    nlp = spacy.blank(lang)
    tagger = nlp.create_pipe('tagger')
    print(TAG_MAP.items())
    for tag, values in TAG_MAP.items():
        print(tag)
        print(values["pos"])
        tagger.add_label(tag, values)

    nlp.add_pipe(tagger)
    optimizer = nlp.begin_training()
    for i in range(n_iter):
        random.shuffle(TRAIN_DATA)
        losses = {}
        for text, annotations in TRAIN_DATA:
            nlp.update([text], [annotations], sgd=optimizer, losses=losses)
        print(losses)
    test_text = "I like blue eggs"
    doc = nlp(test_text)
    print('Tags', [(t.text, t.tag_, t.pos_) for t in doc])

    if output_dir is not None:
        output_dir = Path(output_dir)
        if not output_dir.exists():
            output_dir.mkdir()
        nlp.to_disk(output_dir)
        print("Saved model to", output_dir)

        # test the save model
        print("Loading from", output_dir)
        nlp2 = spacy.load(output_dir)
        doc = nlp2(test_text)
        print('Tags', [(t.text, t.tag_, t.pos_) for t in doc])

if __name__ == '__main__':
    plac.call(main)

# Expected output:
# [
#   ('I', 'N', 'NOUN'),
#   ('like', 'V', 'VERB'),
#   ('blue', 'J', 'ADJ'),
#   ('eggs', 'N', 'NOUN')
# ]

It fails with the following output and traceback. :
=====================================================
dict_items([('N', {'pos': 'NOUN'}), ('V', {'pos': 'VERB'}), ('J', {'pos': 'ADJ'})])
N
NOUN

Traceback (most recent call last):
  File "/home/essin/nlp/train/main.py", line 64, in <module>
    plac.call(main)
    
  File "/home/essin/anaconda3/envs/nlp/lib/python3.9/site-packages/plac_core.py", line 348, in call
    cmd, result = parser.consume(arglist)
    
  File "/home/essin/anaconda3/envs/nlp/lib/python3.9/site-packages/plac_core.py", line 217, in consume
    return cmd, self.func(*(args + varargs + extraopts), **kwargs)
    
  File "/home/essin/nlp/train/main.py", line 36, in main
    tagger.add_label(tag, values)
    
  File "spacy/pipeline/tagger.pyx", line 276, in spacy.pipeline.tagger.Tagger.add_label
TypeError: add_label() takes exactly 2 positional arguments (3 given)
=====================================================

**Comment by ljvmiranda921** [comment, marked as answer]
Hi @Dan-Essin ! Will definitely appreciate it if you [format the code block properly](https://docs.github.com/en/github/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks) so that it's easier to debug.  🙇 

It seems like you're following a book that uses the [v2 API](https://v2.spacy.io/api/tagger#add_label) but the spaCy version you have in your computer is already [v3](https://spacy.io/api/tagger#add_label). The function parameters have changed during that version bump. To confirm which spaCy version you have, try running the following:

```sh
python -m spacy info
```

You have two options here:
- **Install a v2 version of spaCy and rerun the code**: find out what spaCy version the book follows and install that instead
- **Port your code to the v3 version**: the Tagger API has changed, `add_label` doesn't accept a `values` param anymore. You can check the [migration doc](https://spacy.io/usage/v3#migrating-training-mappings-exceptions) here, and follow [this discussion here](https://support.prodi.gy/t/where-is-tag-map-gone-in-spacy-3/3693)

---

Lastly, the reason why it says 3 arguments instead of 2 is because Python counts `self` when reporting the error. Because the `add_label` function belongs to an instantiated class, its signature looks like this:

```python
def add_label(self, label):
    # some code
```

So it just expects 2. However, we are passing three arguments in the code: `self` (always happen whenever you instantiate, `label`, and the tag map.
